### PR TITLE
Add roles

### PR DESCRIPTION
**Pull request by Senjutsuu**
Provided the possibility to add multiple roles to a user. The roles are now stored in the database.

### CODE REVIEW

**Review by ohvitorino — CHANGES_REQUESTED** [review]
IMO Some minor modifications and it is read to be merge.

File: Entity/UserRole.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+     * @var DateTime
+     *
+     * @ORM\Column(type="datetime")
+     */
+    private $editedOn;
+
+    /**
+     * @param string $roleName
```
Comment (StijnVrolijk) [inline Entity/UserRole.php]: No need to place this PHPDoc here

File: Entity/UserRole.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    public function __construct(string $roleName)
+    {
+        parent::__construct($roleName);
+
+        $this->roleName = $roleName;
+    }
+
+    /**
+     * @return int
```
Comment (StijnVrolijk) [inline Entity/UserRole.php]: No need to place this PHPDoc here, either. You should remove it wherever possible in this file.

File: Entity/UserRole.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+     * @ORM\PreUpdate
+     */
+    public function onPreUpdate(): void
+    {
+        $this->editedOn = new DateTime();
+    }
+
+    public function getCreatedOn(): DateTime
```
Comment (StijnVrolijk) [inline Entity/UserRole.php]: You should group getters whenever possible imo, this should move to the top where the other getters are :)

File: Form/AddBaseUserType.php [hunk 4]
```diff
@@ -38,6 +51,21 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
             [
                 'type' => PasswordType::class,
             ]
+        )->add(
+            'roles',
+            EntityType::class,
+            [
+                'attr' => [
+                    'class' => 'select2',
```
Comment (StijnVrolijk) [inline Form/AddBaseUserType.php]: This seems oddly specific. I'd remove this.

File: Form/EditBaseUserType.php [hunk 5]
```diff
@@ -41,6 +54,21 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
                 'type' => PasswordType::class,
                 'required' => false,
             ]
+        )->add(
+            'roles',
+            EntityType::class,
+            [
+                'attr' => [
+                    'class' => 'select2',
```
Comment (StijnVrolijk) [inline Form/EditBaseUserType.php]: Same here

**Review by StijnVrolijk — CHANGES_REQUESTED** [review]

File: Form/AddBaseUserType.php (hunk 4 above)
Comment (Senjutsumo) [reply Form/AddBaseUserType.php]: We implement the select2 entity bundle in the multi user bundle to have the correct functionality of the user role selection. The other option is to implement it per project but that means we have to set it up each time, so for now we will just implement it in the bundle.

**Review by StijnVrolijk — CHANGES_REQUESTED** [review]

**Review by StijnVrolijk — APPROVED** [review]

**Comment by StijnVrolijk** [comment]
@Senjutsumo can you check the tests?

**Review by ohvitorino — APPROVED** [review]

**Comment by StijnVrolijk** [comment]
@Senjutsumo if you fix the tests we can merge this :)

**Pull request merged**